src/Microsoft.AspNetCore.SignalR.Core/HubConnectionContext.cs

```diff
@@ -428,24 +433,48 @@ internal Task AbortAsync()
 
         private void KeepAliveTick()
         {
-            var timestamp = Stopwatch.GetTimestamp();
+            var currentTime = DateTime.UtcNow.Ticks;
             // Implements the keep-alive tick behavior
             // Each tick, we check if the time since the last send is larger than the keep alive duration (in ticks).
             // If it is, we send a ping frame, if not, we no-op on this tick. This means that in the worst case, the
             // true "ping rate" of the server could be (_hubOptions.KeepAliveInterval + HubEndPoint.KeepAliveTimerInterval),
             // because if the interval elapses right after the last tick of this timer, it won't be detected until the next tick.
-            if (timestamp - Interlocked.Read(ref _lastSendTimestamp) > _keepAliveDuration)
+            if (currentTime - Interlocked.Read(ref _lastSendTimestamp) > _keepAliveInterval.Ticks)
             {
                 // Haven't sent a message for the entire keep-alive duration, so send a ping.
                 // If the transport channel is full, this will fail, but that's OK because
                 // adding a Ping message when the transport is full is unnecessary since the
                 // transport is still in the process of sending frames.
                 _ = TryWritePingAsync();
 
-                Interlocked.Exchange(ref _lastSendTimestamp, timestamp);
+                Interlocked.Exchange(ref _lastSendTimestamp, currentTime);
             }
         }
 
+        public void StartPeriodicallyCheckingForClientTimeout()
+        {
+            if (_clientTimeoutActive)
+            {
+                return;
+            }
+            _clientTimeoutActive = true;
+            Features.Get<IConnectionHeartbeatFeature>()?.OnHeartbeat(state => ((HubConnectionContext)state).CheckClientTimeout(), this);
+        }
+
+        private void CheckClientTimeout()
+        {
+            // if it's been too long since we've heard from the client, then close this
+            if (DateTime.UtcNow.Ticks - Interlocked.Read(ref _lastReceivedTimestamp) > _clientTimeoutInterval.Ticks)
+            {
+                Abort();
+            }
+        }
+
+        public void ResetClientTimeout()
+        {
+            _lastReceivedTimestamp = DateTime.UtcNow.Ticks;
```

Comment:
Some places the ticks are being called time and others it is timestamp. I don't have a preference, just use one consistently.